### PR TITLE
fix 4340 bug：duplicate extension

### LINKED ISSUE #4340: ExtensionLoader 加载重复 filter，出现 Duplicate extension 被 exceptions 获取后却不抛出来
- [ ] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have checked the [FAQ](https://github.com/apache/dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 2.6.5
* Operating System version: win10
* Java version: 1.8

### Steps to reproduce this issue

1. dubbo spi 扩展列表有 exception=com.alibaba.dubbo.rpc.filter.ExceptionFilter
2. 我自己实现了一个 filter 扩展，名为 exception=xxxFilter
3. debug com.alibaba.dubbo.common.extension.ExtensionLoader#loadClass 将抛出 throw new IllegalStateException("Duplicate extension "......)

### Expected Result

此异常被 exceptions 获取后，不抛出来，理论上不应该抛出来给用户吗，否则将被加载 jar 包的顺序影响，如果在 dubbo 生效之前加载，自己的 filter 就起作用，否则不起作用

### Actual Result

捕获异常后没有任何提示信息

### PR DESCRIPTION
**Pull request by masteryourself**
## What is the purpose of the change

fix #4340 bug：duplicate extension

## Brief changelog

add duplicateExceptionMsg variable to log duplicateException‘  message

## Verifying this change

run org.apache.dubbo.common.extension.ExtensionLoaderTest#test_LoadExtension_Duplicate

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java [hunk 1]
```diff
@@ -765,7 +770,15 @@ private void saveInExtensionClass(Map<String, Class<?>> extensionClasses, Class<
...
-            throw new IllegalStateException("Duplicate extension " + type.getName() + " name " + name + " on " + c.getName() + " and " + clazz.getName());
+            duplicateExceptionMsg.append("Duplicate extension ")
+                    .append(type.getName())
+                    .append(" name ")
+                    .append(name)
+                    .append(" on ")
+                    .append(c.getName())
+                    .append(" and ")
+                    .append(clazz.getName())
+                    .append("\r\n");
```
Comment (manzhizhen) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java]: I recommend switching to System.lineSeparator() better, what do you think?

**Comment by manzhizhen** [comment]
I will see it again today, I feel that it is a bit strange to change the law.

**Comment by manzhizhen** [comment]
Hello, I looked at your PR. The basic idea is to throw an exception when loading an extension point with the same name but different types, but I still have some objections to this PR. First, you will use it. What is the extension point name already used inside a dubbo? Second, if you have a conflicting extension point name, the class loading order cannot be determined, so it may be loaded into your own Filter, but is it really necessary to throw an exception? For example, when the same class loader loads a class, it is mainly loaded first. Even if there is another identical class, it is ignored by default. Do you really need to throw an exception?

你好，我看了下你的PR，基本的思路是希望当加载到相同名称但类型不同的扩展点时，抛异常出来，但我对这个PR还是有些异议，第一，为啥你会用一个dubbo内部已经使用的扩展点名称？第二，假如你起了一个冲突的扩展点名称，类加载顺序不能确定，所以有可能是加载到你自己的Filter，但是否真需要抛异常吗？比如同一个类加载器加载类时，以先加载的为主，哪怕有另一个相同的class，默认它也是忽略，真的需要抛异常吗？

如果有时间我们可以电话沟通下：18668225833

**Comment by masteryourself** [comment]
我是这么想的，首先这个异常被 catch 后，是会存入 exceptions 里，但是这个信息并没有被日志打出，我觉得至少得给个 warn 级别日志。参考 spring 容器设计，加载重复的 bean 是会抛出异常的，假设说我对dubbo 内部的 filter 扩展点不是很熟悉的情况下，我写了一个重复 key 的 filter，是会有受加载顺序影响的，我觉得这个异常比较重要，应该抛出来。如果不抛出来，同样的代码可能会因为 jar 包的顺序导致容器的 filter 不一样

**Comment by manzhizhen** [comment]
> 我是这么想的，首先这个异常被 catch 后，是会存入 exceptions 里，但是这个信息并没有被日志打出，我觉得至少得给个 warn 级别日志。参考 spring 容器设计，加载重复的 bean 是会抛出异常的，假设说我对dubbo 内部的 filter 扩展点不是很熟悉的情况下，我写了一个重复 key 的 filter，是会有受加载顺序影响的，我觉得这个异常比较重要，应该抛出来。如果不抛出来，同样的代码可能会因为 jar 包的顺序导致容器的 filter 不一样



> 我是这么想的，首先这个异常被 catch 后，是会存入 exceptions 里，但是这个信息并没有被日志打出，我觉得至少得给个 warn 级别日志。参考 spring 容器设计，加载重复的 bean 是会抛出异常的，假设说我对dubbo 内部的 filter 扩展点不是很熟悉的情况下，我写了一个重复 key 的 filter，是会有受加载顺序影响的，我觉得这个异常比较重要，应该抛出来。如果不抛出来，同样的代码可能会因为 jar 包的顺序导致容器的 filter 不一样

其实你的想法是好的，如果冲突，那么就抛异常出来，但目前ExtensionLoader的设计是尽可能的在无法使用时才抛异常（比如类不存在），而对于key重复，因为总会先拿到一个key对应的扩展点，所以它认为是能run起来的。如果真要做到严格校验（失败就终止初始化），那就不能这样改，应该把ExtensionLoader#exceptions这个机制去掉，你觉得呢？

**Comment by manzhizhen** [comment]
直接加微信聊吧，我们看看能否改彻底点（18668225833）

**Comment by beiwei30** [comment]
> 我是这么想的，首先这个异常被 catch 后，是会存入 exceptions 里，但是这个信息并没有被日志打出，我觉得至少得给个 warn 级别日志。参考 spring 容器设计，加载重复的 bean 是会抛出异常的，假设说我对dubbo 内部的 filter 扩展点不是很熟悉的情况下，我写了一个重复 key 的 filter，是会有受加载顺序影响的，我觉得这个异常比较重要，应该抛出来。如果不抛出来，同样的代码可能会因为 jar 包的顺序导致容器的 filter 不一样

I agree we should throw it, since if we don't throw directly then we have no chance to notice user, or at least we should print a warning log.

**Comment by masteryourself** [comment]
nobody review? @manzhizhen

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ApplicationConfig.java [hunk 2]
```diff
@@ -147,6 +148,11 @@
      */
     private String shutwait;
 
+    /**
+     * when in initialization process,throw {@link org.apache.dubbo.common.extension.DubboInitCheckException},stop program
+     */
+    private Boolean check;
```
Comment (chickenlj) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ApplicationConfig.java]: I think `dubbo.application.check` is a little too ambiguous.

Would `extension-loader.duplicate.check` be better? And I also think there's no necessity to declare it in the main trunk, users who want to change the default value can declare in `parameters`.
1. XML
```xml
<dubbo:application>
   <dubbo:parameter key="dubbo.application.extension-loader.duplicate.check" value=true />
</dubbo:application>
```

2. properties file
```properties
dubbo.application.parameters.extension-loader.duplicate.check=true
```

Comment (masteryourself) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ApplicationConfig.java]: Maybe our original intention is different. I assume a scene: when I accidentally customize a dubbo internal filter（just the same name） in another framework, when user use these two jars(our framework & dubbo), how can you guarantee who is loaded first. According to dubbo currently, the first filter loaded will enter the filter chain, and the later loaded same name filter will throw an exception, but this exception is eaten, how do you notify the user. At least we have encountered this situation in Win10 and Centos7, and there is no exception, but the filter does not work.

**Pull request closed without merging**